crates/sui-network/src/state_sync.rs

```diff
@@ -0,0 +1,1433 @@
+        if let Some(tx) = tx.upgrade() {
+            let _ = tx.send(StateSyncMessage::StartSyncJob).await;
+        }
+    }
+}
+
+fn update_peer_height(
+    peer_heights: &RwLock<PeerHeights>,
+    peer_id: PeerId,
+    response: &Result<Option<Checkpoint>, Status>,
+) {
+    match response {
+        Ok(latest) => {
+            peer_heights
+                .write()
+                .unwrap()
+                .update_peer_height(peer_id, latest.clone());
+        }
+        Err(status) => {
+            // TODO maybe don't clear on timeout?
+            trace!("get_latest_checkpoint_summary request failed: {status:?}");
+            peer_heights.write().unwrap().heights.remove(&peer_id);
+        }
+    }
+}
+
+async fn handle_tick(
+    network: anemo::Network,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    tx: mpsc::WeakSender<StateSyncMessage>,
+) {
+    let peer_heights = &peer_heights;
+    let futs = peer_heights
+        .read()
+        .unwrap()
+        .heights
+        .keys()
+        .flat_map(|peer_id| network.peer(*peer_id))
+        .map(|peer| {
+            let peer_id = peer.peer_id();
+            let mut client = StateSyncClient::new(peer);
+
+            let request =
+                Request::new(GetCheckpointSummaryRequest::Latest).with_timeout(DEFAULT_TIMEOUT);
+            async move {
+                let response = client
+                    .get_checkpoint_summary(request)
+                    .await
+                    .map(Response::into_inner);
+                update_peer_height(peer_heights, peer_id, &response);
+                response
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let checkpoints = futures::future::join_all(futs)
+        .await
+        .into_iter()
+        .flatten()
+        .flatten();
+
+    let highest_checkpoint = checkpoints.max_by_key(|checkpoint| checkpoint.sequence_number());
+
+    let our_heighest_checkpoint = peer_heights
+        .read()
+        .unwrap()
+        .highest_known_checkpoint()
+        .cloned();
+
+    let _new_checkpoint = match (highest_checkpoint, our_heighest_checkpoint) {
+        (Some(theirs), None) => theirs,
+        (Some(theirs), Some(ours)) if theirs.sequence_number() > ours.sequence_number() => theirs,
+        _ => return,
+    };
+
+    if let Some(tx) = tx.upgrade() {
+        let _ = tx.send(StateSyncMessage::StartSyncJob).await;
+    }
+}
+
+async fn sync_to_checkpoint<S: CheckpointStore>(
+    network: anemo::Network,
+    store: S,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    checkpoint_event_sender: broadcast::Sender<VerifiedCheckpoint>,
+    checkpoint: Checkpoint,
+) -> Result<VerifiedCheckpoint> {
+    use rand::seq::SliceRandom;
+
+    let mut current = store.get_latest_checkpoint();
+    if current.as_ref().map(|x| x.sequence_number()) >= Some(checkpoint.sequence_number()) {
+        return Err(anyhow::anyhow!(
+            "why are we trying to sync an older checkpoint?"
```

Comment:
this is cute but i have mixed feelings on errors worded as frustrated questions :)

whether or not that part stays - maybe add the current and new checkpoint nums to the message?